### ISSUE TITLE
Deadlock on 32-bit Big-Endian ARM platforms in multi-threaded applications

### ISSUE THREAD

**Issue by slydder101** [libssh2/libssh2#447]
**Describe the bug**
Random deadlocks and crashes appear when using `libssh2+openSSL` or `libssh2+mbedTLS` with `ECDH` key exchange on ` Intel Puma 6 family SoC` 23 Big-Endian platform.
In order to have ECDH support we have applied the path from [PR-385](https://github.com/libssh2/libssh2/pull/385).
A similar issue was previously reported for `libssh2+openSSL` and closed without any solutions [BUG-406](https://github.com/libssh2/libssh2/issues/406).
The issue is not reproducible under `x86_64` platforms.

**To Reproduce**
The bug is almost always reproducible on the specified platform using the following example:
https://gist.github.com/IoanCosminSzanto/de762f982c1acf3afb47f0e2905cd060

When crash or deadlock appears the one of the following callstacks can be seen:

```
deadlock 1

#0  0x0004283a in __pthread_mutex_unlock_usercnt ()
#1  0x000428cc in pthread_mutex_unlock ()
#2  0x0003f47c in _pthread_cleanup_pop_restore ()
#3  0x0003ee46 in calloc ()
#4  0x0001dc50 in mbedtls_mpi_grow ()
#5  0x0001dd82 in mbedtls_mpi_copy ()
#6  0x0001e692 in mbedtls_mpi_sub_abs ()
#7  0x0001e6ec in mbedtls_mpi_add_mpi ()
#8  0x00024276 in ecp_double_jac ()
#9  0x00024dac in ecp_mul_comb ()
#10 0x000257f0 in mbedtls_ecp_mul ()
#11 0x00025db4 in mbedtls_ecp_mul_shortcuts ()
#12 0x00025e12 in mbedtls_ecp_muladd ()
#13 0x00023ac0 in mbedtls_ecdsa_verify ()
#14 0x000170c2 in _libssh2_mbedtls_ecdsa_verify ()
#15 0x0001882e in hostkey_method_ssh_ecdsa_sig_verify ()
#16 0x00019ad6 in ecdh_sha2_nistp ()
#17 0x0001aad8 in kex_method_ecdh_key_exchange ()
#18 0x0001c39c in _libssh2_kex_exchange ()
#19 0x00013b8e in libssh2_session_handshake ()
#20 0x000102d6 in request () at ssh2_stress.c:45
#21 0x00010474 in do_request (data=0x0) at ssh2_stress.c:92
#22 0x000435b2 in start_thread ()
#23 0x000444dc in clone ()

deadlock 2

#0  0x00041f62 in __lll_lock_wait ()
#1  0x000424ba in pthread_mutex_lock ()
#2  0x0003ee18 in calloc ()
#3  0x0001dc50 in mbedtls_mpi_grow ()
#4  0x0001dd82 in mbedtls_mpi_copy ()
#5  0x0001e692 in mbedtls_mpi_sub_abs ()
#6  0x00027310 in ecp_mod_p256 ()
#7  0x00023d4e in ecp_modp ()
#8  0x00024980 in ecp_add_mixed ()
#9  0x000251a2 in ecp_mul_comb ()
#10 0x000257f0 in mbedtls_ecp_mul ()
#11 0x000233b6 in mbedtls_ecdh_compute_shared ()
#12 0x00017026 in _libssh2_mbedtls_ecdh_gen_k ()
#13 0x00019912 in ecdh_sha2_nistp ()
#14 0x0001aad8 in kex_method_ecdh_key_exchange ()
#15 0x0001c39c in _libssh2_kex_exchange ()
#16 0x00013b8e in libssh2_session_handshake ()
#17 0x000102d6 in request () at ssh2_stress.c:45
#18 0x00010474 in do_request (data=0x0) at ssh2_stress.c:92
#19 0x000435b2 in start_thread ()
#20 0x000444dc in clone ()

crash 1:

#0  0x000f46b2 in __libc_malloc ()
#1  0x000f4d06 in calloc ()
#2  0x00080768 in mbedtls_mpi_grow ()
#3  0x0008089a in mbedtls_mpi_copy ()
#4  0x00081144 in mbedtls_mpi_sub_abs ()
#5  0x00081254 in mbedtls_mpi_sub_mpi ()
#6  0x00082f06 in mbedtls_mpi_inv_mod ()
#7  0x000871c8 in ecp_normalize_jac ()
#8  0x00088584 in ecp_mul_comb ()
#9  0x00088bac in mbedtls_ecp_mul ()
#10 0x00089170 in mbedtls_ecp_mul_shortcuts ()
#11 0x000891ea in mbedtls_ecp_muladd ()
#12 0x00086e7c in mbedtls_ecdsa_verify ()
#13 0x0006b476 in _libssh2_mbedtls_ecdsa_verify ()
#14 0x00061c4a in hostkey_method_ssh_ecdsa_sig_verify ()
#15 0x0006d936 in ecdh_sha2_nistp ()
#16 0x0006e938 in kex_method_ecdh_key_exchange ()
#17 0x000701fc in _libssh2_kex_exchange ()
#18 0x00064102 in libssh2_session_handshake ()
#19 0x0002cc86 in SSHHidrantCommand::execute (this=0x1d4b08) at Pete2/tasks/vulnerability/Hidr
ant/commands/SSHHidrantCommand.cpp:52

crash 2:

#0  0x000f44ce in __libc_free ()
#1  0x0008073a in mbedtls_mpi_free ()
#2  0x00087a54 in mbedtls_ecp_point_free.part ()
#3  0x00087b86 in mbedtls_ecp_group_free ()
#4  0x00087be2 in mbedtls_ecp_keypair_free ()
#5  0x0008709a in mbedtls_ecdsa_free ()
#6  0x0006b610 in _libssh2_mbedtls_ecdsa_free ()
#7  0x0006e828 in kex_method_ecdh_key_exchange ()
#8  0x000701fc in _libssh2_kex_exchange ()
#9  0x00064102 in libssh2_session_handshake ()
#10 0x0002cc86 in SSHHidrantCommand::execute (this=0x1e3148) at Pete2/tasks/vulnerability/Hidr
ant/commands/SSHHidrantCommand.cpp:52
#11 0x0002e532 in HidrantWorker::start (this=0x19e2e8) at Pete2/tasks/vulnerability/Hidrant/Hi
drantWorker.cpp:38
```

**Version:**
 - OS: Linux
- `libssh2` version 1.9.0
- `openSSL` version 1.1.0.
- `mbedTLS` version 2.7.12

**Additional context**
Some notes based on our investigations:
* The issue appears for bot combinations of `libssh2+openSSL` and `libssh2+mbedTLS(with the applied patch)`
 * The issue was not reproduce when running on a single thread
* The issue is not reproducible if  `libssh2_session_handshake` call is guarded with a mutex
* Using `libssh` with openSSL or mbedTLS doesn't cause deadlock or crash so it is probably not an issue with the the underlying SSL libraries
*  It seams that the issues appears when using `ECDH` keys-exchange, so maybe some initialization is missing when using the SSL libraries in multi-threaded environment?

**Comment by willco007** [comment]
This deadlock is in the mbedtls library.

**Comment by willco007** [comment]
This deadlock and the ones reported in #406 are all outside of libssh2 specifically. It seems something else may be going on with your system libraries outside of libssh2 related to locking and memory allocation.

**Comment by slydder101** [comment]
I don't think the cause of the issue is in the SSL libraries because of the following reasons:
1. the deadlocks both appear with using different implementation of SSL libraries (`mbedTLS` and `openSSL`
2. the bug can not be reproduced with [libssh library](http://www.libssh.org/) on the same machine (tested both combinations of `libssh + mbedTLS` and `libssh + openSSL` with above test application modified to use `libssh` API)
3. the bug does not manifest when using only a single thread
4. running the same binary on `x86_64` architecture with static QEMU emulation exhibits the bug
5. we have other **heavily multi-threaded applications** running on the same machine **built with the absolutely same SDK** (gcc, uClibc, libraries) that work flawlessly 

**Conclusions**:
1. the bug manifests on `ARMv6` architecture (ARMv7+, MIPS, x86 seems to work fine)
2. we think the issue relies in improper use of `mbedTLS` or `openSSL` libraries when multiple threads are used, because both libraries in other context works well on ARMv6
3. it could be some implementation specific parts in libssh2 C code which compiled for ARMv6 results in buggy behavior

Do you have any suggestions how can we help to track down this issue? Would the binary used in reason 4. be helpful to you?

Thank you for your time and insights.

**Comment by willco007** [comment]
To my knowledge there is no ARM specific code in libssh2 and very little big endian specific code. How is your client threaded code written? Are you creating a new session per thread? Are you locking a shared session across threads? Are you certain you've build the encryption libraries correctly for ARMv6?

**Comment by slydder101** [comment]
> How is your client threaded code written? Are you creating a new session per thread?

Yes, that is correct.

> Are you locking a shared session across threads?

No (see the link with the example code in my first post).

> Are you certain you've build the encryption libraries correctly for ARMv6?

Pretty much,  because exactly the same binaries were used in the `libssh + mbedTLS` and `libssh + openSSL` (the very same example modified to use `libssh` API) and it worked without issue.

I'm investigating it further and will let you know if we found something.
Thank you.

**Comment by slydder101** [comment]
Yes, you were right: we confirmed that the issue is outside of `libssh2`. We managed to reproduce the deadlock only with ARMv6 + uClibc-ng with Thumb enabled + NPTL. The deadlock is not reproducible in any other combination (non-Thumb, ARMv7+, glibc, musl). If we compile uClibc-ng with Thumb but LinuxThreads instead of NPTL the deadlock does not appear.

As a conclusion the bug is in the NPTL locking implementation of uClibc-ng compiled for ARMv6 Thumb mode. We have to report it to that project.

Thank you for your time and cooperation, we can close this issue.
